ai-integrations/langgraph.ipynb

```diff
@@ -0,0 +1,598 @@
+    },
+    "id": "cxTXczeTghzU",
+    "outputId": "ae3a81b2-cba6-42fc-f593-8646bff77b14",
+    "vscode": {
+     "languageId": "shellscript"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "pip install --quiet --upgrade langgraph langgraph-checkpoint-mongodb langchain langchain_mongodb langchain-openai pymongo"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "oXLWCWEghuOX"
+   },
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "\n",
+    "os.environ[\"OPENAI_API_KEY\"] = \"<api-key>\"\n",
+    "MONGODB_URI = \"<connection-string>\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "UUf3jtFzO4-V"
+   },
+   "source": [
+    "## Use Atlas as a vector database"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_mongodb import MongoDBAtlasVectorSearch\n",
+    "from langchain_openai import OpenAIEmbeddings\n",
+    "from pymongo import MongoClient\n",
+    "\n",
+    "# Connect to your Atlas cluster\n",
+    "client = MongoClient(MONGODB_URI)\n",
+    "collection = client[\"sample_mflix\"][\"embedded_movies\"]\n",
+    "embedding_model = OpenAIEmbeddings(model=\"text-embedding-ada-002\", disallowed_special=())\n",
+    "\n",
+    "# Instantiate the vector store\n",
+    "vector_store = MongoDBAtlasVectorSearch(\n",
+    "   collection = collection,\n",
+    "   embedding = embedding_model,\n",
+    "   text_key = \"plot\",\n",
+    "   embedding_key = \"plot_embedding\",\n",
+    "   relevance_score_fn = \"dotProduct\"\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Use helper method to create the vector search index\n",
+    "vector_store.create_vector_search_index(\n",
+    "   dimensions = 1536 # The dimensions of the vector embeddings to be indexed \n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_mongodb.index import create_fulltext_search_index\n",
+    "import time\n",
+    "\n",
+    "# Use helper method to create the search index\n",
+    "create_fulltext_search_index(\n",
+    "   collection = collection,\n",
+    "   field = \"title\",\n",
+    "   index_name = \"search_index\"\n",
+    ")\n",
+    "\n",
+    "# Wait for the index to build (this can take around a minute)\n",
+    "time.sleep(60)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "pZfheX5FiIhU"
+   },
+   "source": [
+    "## Define agent tools\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.agents import tool\n",
+    "\n",
+    "# Define a vector search tool\n",
+    "@tool\n",
+    "def vector_search(user_query: str) -> str:\n",
+    "    \"\"\"\n",
```

Comment:
Nit: it's a little strange that we're doing function documentation here with args/returns, when we haven't done that anywhere else in our docs code examples. If we're being consistent with the style of our other comments, we'd probably only keep the first line of this comment.